mypy_django_plugin/transformers/models.py

```diff
@@ -597,6 +609,182 @@ def run_with_model_cls(self, model_cls: Type[Model]) -> None:
+            list(get_field_descriptor_types(default_pk_field, is_set_nullable=True, is_get_nullable=False)),
+        )
+
+    def get_pk_instance(self, model: TypeInfo, /) -> Instance:
+        contains_from_pk_info = model.get_containing_type_info("pk")
+        if contains_from_pk_info is not None:
+            pk = contains_from_pk_info.names["pk"].node
+            if isinstance(pk, Var) and isinstance(pk.type, Instance):
+                return pk.type
+        return self.default_pk_instance
+
+    def resolve_many_to_many_arguments(self, call: CallExpr, /) -> Optional[M2MArguments]:
+        """
+        Inspect a 'ManyToManyField(...)' call to collect argument data on any 'to' and
+        'through' arguments.
+        """
+        look_for: Dict[str, Optional[CallArg]] = {"to": None, "through": None}
+        # 'to' is declared as the first positional argument. But not required positional
+        if call.arg_kinds[0].is_positional():
+            look_for["to"] = CallArg(expr=call.args[0], kind=call.arg_kinds[0], name=call.arg_names[0])
+        # 'through' is declared as the sixth positional argument. But not required
+        # positional
+        if len(call.args) > 5 and call.arg_kinds[5].is_positional():
+            look_for["through"] = CallArg(expr=call.args[5], kind=call.arg_kinds[5], name=call.arg_names[5])
+
+        if any(value is None for value in look_for.values()):
+            for arg_expr, arg_kind, arg_name in zip(call.args, call.arg_kinds, call.arg_names):
+                if arg_name in look_for and look_for[arg_name] is None:
+                    look_for[arg_name] = CallArg(expr=arg_expr, kind=arg_kind, name=arg_name)
+
+        to_arg = look_for["to"]
+        if to_arg is None:
+            # 'ManyToManyField()' requires the 'to' argument
+            return None
+        to_model: Optional[TypeInfo]
+        if isinstance(to_arg.expr, StrExpr) and to_arg.expr.value == "self":
+            to_model = self.model_classdef.info
+            to_self = True
+        else:
+            to_model = get_model_info_from_call_arg(
+                to_arg,
+                current_module_name=self.model_classdef.info.module_name,
+                api=self.api,
+                django_context=self.django_context,
+            )
+            to_self = False
+        if to_model is None:
+            return None
+        to = M2MTo(arg=to_arg, model=to_model, self=to_self)
+
+        through_arg = look_for["through"]
+        through = None
+        if through_arg is not None:
+            through_model = get_model_info_from_call_arg(
+                through_arg,
+                current_module_name=self.model_classdef.info.module_name,
+                api=self.api,
+                django_context=self.django_context,
+            )
+            if through_model is not None:
+                through = M2MThrough(arg=through_arg, model=through_model)
+
+        return M2MArguments(to=to, through=through)
```

Comment:
This part of the change seems like the most complex one to me.

Reply:
Yeah, it's messy. All it does is sort out how `ManyToManyField()` was called;

- Did it have `to=` and/or `through=`?
- Resolve any lazy reference for `to=` or `through=`
- Include the model (typeinfo) that `to=` and/or `through=` pointed to

I tried to encapsulate these results in the named tuple `M2MArguments`, and also to get some form of declarative description of how we can interpret a call to `ManyToManyField`.

e.g. required `to` but optional `through`, while `through` can also be an implicit model: `M2MThrough(call_arg=None, ...)`